crates/flux-infer/src/lean_format.rs

```diff
@@ -0,0 +1,189 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            Expr::Var(v) => write!(f, "{}", LeanVar(v)),
+            Expr::Constant(c) => {
+                match c {
+                    Constant::Numeral(n) => write!(f, "{}", n),
+                    Constant::Boolean(b) => write!(f, "{}", b),
+                    Constant::String(s) => write!(f, "{}", s.display()),
+                    Constant::Decimal(d) => write!(f, "{}", d.display()),
+                    _ => todo!(),
+                }
+            }
+            Expr::BinaryOp(bin_op, args) => {
+                let bin_op_str = match bin_op {
+                    BinOp::Add => "+",
+                    BinOp::Sub => "-",
+                    BinOp::Mul => "*",
+                    BinOp::Div => "/",
+                    BinOp::Mod => "%",
+                };
+                write!(f, "({} {} {})", LeanExpr(&args[0]), bin_op_str, LeanExpr(&args[1]))
+            }
+            Expr::Atom(bin_rel, args) => {
+                let bin_rel_str = match bin_rel {
+                    BinRel::Eq => "=",
+                    BinRel::Ne => "≠",
+                    BinRel::Le => "≤",
+                    BinRel::Lt => "<",
+                    BinRel::Ge => "≥",
+                    BinRel::Gt => ">",
+                };
+                write!(f, "({} {} {})", LeanExpr(&args[0]), bin_rel_str, LeanExpr(&args[1]))
+            }
+            Expr::App(function, args) => {
+                write!(
+                    f,
+                    "({} {})",
+                    LeanExpr(function.as_ref()),
+                    args.iter().map(LeanExpr).format(" ")
+                )
+            }
+            Expr::And(exprs) => {
+                write!(f, "({})", exprs.iter().map(LeanExpr).format(" && "))
+            }
+            Expr::Or(exprs) => {
+                write!(f, "({})", exprs.iter().map(LeanExpr).format(" || "))
+            }
+            _ => todo!(),
+        }
+    }
+}
+
+impl<'a> fmt::Display for LeanFunDef<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let LeanFunDef(FunDef { name, args, out, comment: _, body }) = self;
+        writeln!(
+            f,
+            "def {} {} : {} :=",
+            LeanVar(name),
+            args.iter()
+                .map(|(arg, arg_sort)| format!("({} : {})", LeanVar(arg), LeanSort(arg_sort)))
+                .collect::<Vec<_>>()
+                .join(" "),
+            LeanSort(out)
+        )?;
+        writeln!(f, "  {}", LeanExpr(body))
+    }
+}
+
+impl<'a> fmt::Display for LeanPred<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            Pred::Expr(expr) => write!(f, "{}", LeanExpr(expr)),
+            Pred::And(preds) => write!(f, "({})", preds.iter().map(LeanPred).format(" ∧ ")),
+            Pred::KVar(_, _) => panic!("kvars should not appear when encoding in lean"),
```

Comment:
Just a note to myself. I'm not suggesting we implemented this, but we could actually support kvars by existentially quantifying over them